File: driver.js
// Fluoresce DB "driver"
const net = require('net');
const PassKey = "";

function Perform(ToSend) {
	ToSend['passkey'] = PassKey;
	return new Promise((resolve, reject) => {
		let Response = "";
		const socket = net.connect(4781, "127.0.0.1");
		socket.end(JSON.stringify(ToSend));
		socket.on("data", (data) => {
			Response += data;
		});
		socket.on("end", () => {
			socket.destroy();
			Response = JSON.parse(Response);
			if (Response['error'] != 0) {
				let Line = "Fluoresce: ";
				switch(Response['error']) {
					case 1:
						Line += "The provided passkey is incorrect.";
						break;
					case 2:
						Line += "The \"confirmation\" value is required for dangerous operations.";
						break;
					case 3:
						Line += "A User ID value is required for Read/Write/Delete.";
						break;
					case 5:
						Line += "Data required for write.";
						break;
					case 99:
						Line += "The data sent was not valid JSON.";
						break;
				}
				console.log(Line);
				resolve(Response['error']);
			}
			else {
				resolve(Response['data']);
			}
		});
	});
}

function Create(Database, Type, Data, UserID) {
	return Perform({
		'type': "create",
		'db': Database,
		'createtype': Type != undefined ? Type : false,
		'data': Data != undefined ? Data : false,
		'userid': UserID != undefined ? UserID : false
	});
}
function Destroy(Database, UserID, Confirmation) {
	return Perform({
		'type': "destroy",
		'db': Database,
		'userid': UserID != undefined ? UserID : false,
		'confirmation': Confirmation != undefined ? Confirmation : false
	});
}

function Read(Database, UserID, Target, Search) {
	return Perform({
		'type': "read",
		'db': Database,
		'userid': UserID,
		'target': Target != undefined ? Target : false,
		'search': Search != undefined ? Search : false
	});
}
function DirectRead(Database, UserID) {
	return Perform({
		'type': "directread",
		'db': Database,
		'userid': UserID != undefined ? UserID : false
	});
}
function ReadIncrement(Database, UserID) {
	return Perform({
		'type': "readincrement",
		'db': Database,
		'userid': UserID != undefined ? UserID : false
	});
}

function Write(Database, UserID, Data, Target, Search) {
	return Perform({
		'type': "write",
		'db': Database,
		'userid': UserID,
		'data': Data,
		'target': Target != undefined ? Target : false,
		'search': Search != undefined ? Search : false,
	});
}
function DirectWrite(Database, UserID, Data) {
	return Perform({
		'type': "directwrite",
		'db': Database,
		'userid': UserID,
		'data': Data
	});
}

function Delete(Database, UserID, Target, Search, Confirmation) {
	return Perform({
		'type': "delete",
		'db': Database,
		'userid': UserID,
		'target': Target != undefined ? Target : false,
		'search': Search != undefined ? Search : false,
		'confirmation': Confirmation != undefined ? Confirmation : false
	});
}

function ReadList(Database, UserID, Count, KeyID, KeyValue) {
	return Perform({
		'type': "readlist",
		'db': Database,
		'userid': UserID,
		'count': Count != undefined ? Count : false,
		'keyname': KeyID != undefined ? KeyID : false,
		'keyvalue': KeyValue != undefined ? KeyValue : false
	});
}
function Append(Database, UserID, Data) {
	return Perform({
		'type': "append",
		'db': Database,
		'userid': UserID,
		'data': Data
	});
}
function QueryList(Database, UserID, Search) {
	return Perform({
		'type': "querylist",
		'db': Database,
		'userid': UserID,
		'search': Search
	});
}

function Exists(Database, UserID) {
	return Perform({
		'type': "exists",
		'db': Database,
		'userid': UserID != undefined ? UserID : false
	});
}

function List(Database) {
	return Perform({
		'type': "list",
		'db': Database
	});
}

function Count(Database, UserID, Target) {
	return Perform({
		'type': "count",
		'db': Database,
		'userid': UserID != undefined ? UserID: false,
		'target': Target != undefined ? Target : false
	});
}

function CountUser(Database, UserID, Target) {
	return Perform({
		'type': "countusers",
		'db': Database,
		'userid': UserID != undefined ? UserID: false,
		'target': Target != undefined ? Target : false
	});
}

function Save(Database) {
	return Perform({
		'type': "save",
		'db': Database
	});
}

module.exports = { Create, Destroy,
				   Read, DirectRead, ReadIncrement,
				   Write, DirectWrite,
				   Delete,
				   ReadList, QueryList, Append,
				   Exists, List,
				   Count, CountUser, Save };

File: driver.lua
local coro_net = require('coro-net');
local JSON = require('json');

local Module = {}

function Module.Create(Database)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "create", destination = Database });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.Delete(Database, UserID)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "delete", destination = Database, userid = UserID });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.Read(Database, UserID)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "read", destination = Database, userid = UserID });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.Write(Database, UserID, Data)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "write", destination = Database, userid = UserID, data = Data });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.DirectRead(Database, UserID)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "directread", destination = Database, userid = UserID });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.DirectWrite(Database, UserID, Data)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "directwrite", destination = Database, userid = UserID, data = Data });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.ReadIndex(Database, UserID, Index)
	-- Index: {'valuename': "", 'value': ?} --
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "readindex", destination = Database, userid = UserID, index = Index });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.WriteIndex(Database, UserID, Index, Data)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "writeindex", destination = Database, userid = UserID, index = Index, data = Data });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.DeleteIndex(Database, UserID, Index)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "deleteindex", destination = Database, userid = UserID, index = Index });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.ReadObject(Database, UserID, ObjectName)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "readobject", destination = Database, userid = UserID, objectname = ObjectName });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.WriteObject(Database, UserID, ObjectName, Data)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "writeobject", destination = Database, userid = UserID, objectname = ObjectName, data = Data });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.DeleteObject(Database, UserID, ObjectName)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "deleteobject", destination = Database, userid = UserID, objectname = ObjectName });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.ReadObjectIndex(Database, UserID, ObjectName, Index)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "readobjectindex", destination = Database, userid = UserID, objectname = ObjectName, index = Index });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.WriteObjectIndex(Database, UserID, ObjectName, Index, Data)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "writeobjectindex", destination = Database, userid = UserID, objectname = ObjectName, index = Index, data = Data });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.DeleteObjectIndex(Database, UserID, ObjectName, Index)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "deleteobjectindex", destination = Database, userid = UserID, objectname = ObjectName, index = Index });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.Exists(Database, UserID)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "exists", destination = Database, userid = UserID });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response['exists'];
end;

function Module.List(Database)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "list", destination = Database });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

function Module.Save(Database)
	local Response = {};
	local read, write = coro_net.connect({
		host = "127.0.0.1",
		port = 4781
	});
	local Request = JSON.stringify({ type = "forcesave", destination = Database });
	write(Request); write();
	for Reply in read do
		Response[#Response + 1] = Reply;
	end;
	Response = table.concat(Response);
	return Response;
end;

return Module;
